### ISSUE TITLE
Use entire device as keyfile

### ISSUE THREAD

**Issue by giddie** [johnlane/grub#7]
I like to have some two-factor security by keeping my keyfile password-protected on my USB key by storing it in its own LUKS container (to keep code complexity at a minimum). This works great in my initrd, basically doing this:

```
mount /dev/usb /mnt
cryptsetup luksOpen /mnt/keyfile.luks keyfile  # prompts for password
cryptsetup luksOpen --key-file /dev/mapper/keyfile /dev/crypted-device decrypted
root=/dev/mapper/decrypted
```

I'd really like to push this scheme up to the GRUB layer. Using your patches, I'm able to read a keyfile to open my crypted devices, but can't quite get my scheme working because (using keyfile.luks on memdisk for testing:)

```
grub> ls (memdisk)/
keyfile.luks
grub> loopback keydev (memdisk)/keyfile.luks
grub> ls
(keydev) (memdisk) ...
grub> cryptomount (keydev)
Attempting to decrypt master key...
Enter passphrase for keydev (...):
Slot 0 opened
grub> ls
(keydev) (memdisk) (crypto0) ...
grub> cryptomount -u <crypted-device-uuid> -k (crypto0)   # No good syntax for this?
Unable to open key file (crypto0)
error: disk `crypto0' not found.
```

I'd rather not have to embed a filesystem (or tar archive) in my keyfile.luks container, so I wonder if an option could be added to your patch in order to make it possible to read the key from a device as well as a file?

**Comment by johnlane** [comment]
The ke file is opened by a Grub routine called `grub_file_open`. The relevant code is in my [second patch](http://grub.johnlane.ie/assets/0002-Cryptomount-support-key-files.patch) and is basically this:

```
  keyfile = grub_file_open (state[4].arg);
```

where `state[4].arg` is whatever you give as a parameter to `-k`.

Now, I don't know whether `grub_file_open` will open a device and, if it would, I don't know how it would expect that to be presented. I would suspect it is as you have done and would deduce from your difficulty that it doesn't support it.

If you want to look at it yourself and submit a PR if you are successful then I wouldn't mind incorporating such an enhancement here. I'm a bit reluctant to spend more time on it myself because upstream have shown little interest in adopting this.

---

Having taken a quick look at the Grub source in [file.c](http://git.savannah.gnu.org/cgit/grub.git/tree/grub-core/kern/file.c) I would think adding another option to do a `grub_device_open` might be a possibility but the library isn't documented (as far as I am aware) so it would be a case of experimenting, trial and error. If you think you can manage it then by all means give it a go...

**Comment by giddie** [comment]
I've given this some thought, and it seems to me that since GRUB has a strong concept of a device vs a file, it might become somewhat ugly to introduce the device-as-file concept at a low level. In your opinion, how difficult would it be to create a new module, similar to the `loopback', but the opposite way around. For instance:

```
grub> rawdevice keyfile_raw (crypto0)
grub> ls
(keydev) (memdisk) (crypto0) (keyfile_raw) ...
grub> ls (keyfile_raw)/
fulldevice
```

With this API, it should also be possible to implement offset + length and similar tweaks directly in the filename, though that's not something I particularly need right now. Any thoughts regarding this, or the difficulty of implementation?

**Comment by giddie** [comment]
Sorry, I misread your comment regarding `grub_device_open`. I thought you were suggesting changing a low-level library. Having taken a closer look at the code, it doesn't look overly difficult to add another option to read from a device instead of a file. Although I still like the elegance of having a separate module to map a device to a file, it would probably be an unreasonable amount of work just to get this feature working :p

I'll see if I can get some code together.

**Comment by johnlane** [comment]
Merged #8 
I have also updated the web documentation to refer to the new feature.
